request.d.ts: Options union types

In request.d.ts ([lines](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4ec3c5bf291bc02f49eb30c1077340b235165c67/request/request.d.ts#L116-L122)) it defines options as union types:

``` js
interface CoreOptions {
    // ...
}
interface UriOptions {
    uri: string;
}
interface UrlOptions {
    url: string;
}
export type RequiredUriUrl = UriOptions | UrlOptions;
export type Options = RequiredUriUrl & CoreOptions;
```

With this definition, it's unable to access (read or overwrite) the `uri`/`url` property. Below TypeScript code fail to compile:

``` js
const opt: Options = {
  baseUrl: 'http://localhost',
  uri: 'bar'
};

console.log(opt.uri);
// error TS2339: Property 'uri' does not exist on type 'UriOptions | UrlOptions'.
```

--- a/request/request-tests.ts
+++ b/request/request-tests.ts
@@ -110,6 +110,15 @@ var options: request.Options = {
 	strictSSL: bool
 };
 
+// Below line has compile error, use OptionsWithUri or OptionsWithUrl instead. See #7979.
+// options.uri = str;
+
+const opt: request.OptionsWithUri = {
+  baseUrl: 'http://localhost',
+  uri: 'bar'
+};
+opt.uri = str;
+
 // --- --- --- --- --- --- --- --- --- --- --- ---
 
 agent = req.getAgent();

--- a/request/request.d.ts
+++ b/request/request.d.ts
@@ -125,7 +125,10 @@ declare module 'request' {
 			uri?: string;
 			url?: string;
 		}
-		export type Options = RequiredUriUrl & CoreOptions;
+
+        export type OptionsWithUri = UriOptions & CoreOptions;
+        export type OptionsWithUrl = UrlOptions & CoreOptions;
+        export type Options = OptionsWithUri | OptionsWithUrl;
 
 		export interface RequestCallback {
 			(error: any, response: http.IncomingMessage, body: any): void;

